_posts/2021-05-18-my-first-steps-in-opensearch-plugins.markdown

```diff
@@ -0,0 +1,619 @@
+    `-- yamlRestTest
+        |-- java
+        |   `-- org
+        |       `-- opensearch
+        |           `-- rest
+        |               `-- action
+        |                   `-- HelloWorldClientYamlTestSuiteIT.java
+        `-- resources
+            `-- rest-api-spec
+                |-- api
+                |   `-- _plugins.hello_world.json
+                `-- test
+                    `-- _plugins.hello_world
+                        |-- 10_basic.yml
+                        `-- 20_hello_world.yml
+```
+
+---
+## The `build.gradle` of Our Plugin
+All OpenSearch plugins have the same general structure.
+The plugin itself is a zip file that contains at the root:
+- The plugin `jar` files
+- `plugin-descriptor.properties`
+- [OPTIONAL] `bin` directory
+- [OPTIONAL] `config` directory 
+- [OPTIONAL] `plugin-security.policy`
+
+> Any resources a plugin requires are packaged into a resources `jar`. This is due to the fact that only `jar` files at the root are added to the classpath for the plugin.
+
+The `build.gradle` file will build this structure for us when we are done, including creating the required `plugin-descriptor.properties`.
+
+However, since we need the OpenSearch project artifact for the gradle build process and the plugin interface, let's clone the project and publish it to our Maven local. This last step is required since the artifact is not yet published to public repositories. 
+
+In this example I will be using release version `1.0.0-beta1`, as described [here](https://github.com/opensearch-project/opensearch-plugins/blob/main/README.md):
+
+```
+➜  ~ git clone https://github.com/opensearch-project/OpenSearch.git
+➜  ~ cd OpenSearch
+➜  OpenSearch git:(main) git checkout 1.0.0-beta1 -b beta1-release
+➜  OpenSearch git:(beta1-release) ./gradlew publishToMavenLocal -Dbuild.version_qualifier=beta1 -Dbuild.snapshot=false
+```
+
+
+Copy the following to the `build.gradle` file:
+```
+apply plugin: 'java'
+apply plugin: 'opensearch.opensearchplugin'
+apply plugin: 'opensearch.yaml-rest-test'
+
+opensearchplugin {
+    name 'opensearch-rest-plugin'
+    description 'Custom OpenSearch REST plugin for educational purposes'
+    classname 'org.opensearch.rest.action.HelloWorldPlugin'
+    licenseFile rootProject.file('LICENSE.txt')
+    noticeFile rootProject.file('NOTICE.txt')
+}
+
+buildscript {
+    repositories {
+        mavenCentral()
+        jcenter()
+        mavenLocal()
+    }
+
+    dependencies {
+        classpath "org.opensearch.gradle:build-tools:1.0.0-beta1"
+    }
+}
+
+repositories {
+    mavenLocal()
+}
+
+// disabling some unnecessary validations for this plugin
+testingConventions.enabled = false
+loggerUsageCheck.enabled = false
+validateNebulaPom.enabled = false
+
+dependencies {
+    // required for the yaml test to run
+    yamlRestTestImplementation 'org.apache.logging.log4j:log4j-core:2.11.1'
+}
+```
+
+A few things to note on the `build.gradle`:
+1. `opensearchplugin` section:
```

Comment:
@saratvemulapalli this section explains in short terms the descriptor file and its relation to the `opensearchplugin` in the `build.gradle`.

Reply:
Thank you Amitai! 
My bad long changes, didn't quite go through till the end.